### PR TITLE
Nix updates 2.3

### PR DESCRIPTION
**Pull request by poelzi**
These is a backport of #3174 to 2.3.

It fixes all issues I found so far in the derivate.
This will allow us to deliver up to date derivates through cachix to the enduser.
This branch now has working travis for nix and runs all tests.

### CODE REVIEW

**Comment by poelzi** [comment]
@haslersn wanna review ?
I'm using this version now for quite some time and it works very nicely. I added nix ci that run the tests and will publish to cachix.
This way you will be able to test a branch without compiling :)

**Comment by Be-ing** [comment]
There are merge conflicts now.

File: CMakeLists.txt
Comment (Be-ing) [inline CMakeLists.txt:1093]: remove this

Comment (poelzi) [reply CMakeLists.txt]: no idea how this got in there

File: .github/workflows/nix.yml
Comment (uklotzde) [inline .github/workflows/nix.yml:10]: Will this add 3 more builds to CI? We can't add builds for every possible distribution out there.

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by Be-ing** [comment]
@poelzi do you still want to merge this for 2.3? If so, please address @uklotzde's review comments above.

**Comment by Holzhaus** [comment]
> @poelzi do you still want to merge this for 2.3? If so, please address @uklotzde's review comments above.

@poelzi Friendly ping.

File: CMakeLists.txt
Comment (Holzhaus) [inline CMakeLists.txt:88]: Why not use `option`? Also, can you add a comment why this can be disabled? And please add that it's gcc/clang only to the help string.

File: .gitignore
Comment (poelzi) [reply .gitignore:67]: I changed it to match to root only

File: CMakeLists.txt
Comment (poelzi) [reply CMakeLists.txt:60]: I don't understand

Comment (Holzhaus) [reply CMakeLists.txt:60]: s/Linux/"Linux"/

**Comment by poelzi** [comment]
@Be-ing any idea why the nixos builds do not show up here anymore ? They worked before but now they do not show up.

**Comment by Holzhaus** [comment]
> @Be-ing any idea why the nixos builds do not show up here anymore ? They worked before but now they do not show up.

```
.github/workflows/nix.yml#L37

The workflow is not valid. .github/workflows/nix.yml (Line: 37, Col: 12): Unexpected symbol: '"false"'. Located at position 21 within expression: matrix.use_clazy || "false"
```

https://github.com/poelzi/mixxx/actions/runs/742358528

File: CMakeLists.txt
Comment (daschuer) [inline CMakeLists.txt:62]: Disabling fast math is a bad idea. I guess the resulting binary is uselessly slow when using EQs.
Can't we just recommend to use gcc?

Comment (poelzi) [reply CMakeLists.txt]: I changed the suggestion to prefer the use gcc and only in extrem cases disable fast math

File: shell.nix
Comment (Holzhaus) [inline shell.nix:56]: why is that in here? Doesn't `pre-commit` work on nkxOS?

File: shell.nix
Comment (Holzhaus) [inline shell.nix:284]: Why are you overriding that variable here?

**Review by uklotzde — COMMENTED** [review]
No strong objections. I cannot comment on the NixOS specific changes.

The FAST_MATH option is useful on its own and could also be added separately. It helps for testing the recently discovered NaN issues.

@poelzi Please respond to or take care of the remaining review comments.

@Holzhaus Your turn.

**Review by uklotzde — APPROVED** [review]

**Comment by Holzhaus** [comment]
> @Holzhaus Your turn.

I also can't comment on the nixOS specific changes. This PR is huge and I still object to the general idea of maintaing nixOS build scripts in our tree when no core dev uses it. It will probably break at some point, and we won't be able to fix it. We cannot and shouldn't add distro-specific scripts for each distro that exists.

Shouldn't this stuff be contributed over here instead? https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/mixxx/default.nix

That said, we already have a broken nix script, so this improves the situation I guess (I cannot tell because I can't test).

I'd like at least a response on the 2 comments before merging though, because I'm worried that this addition clang-format will lead to nixOS users doing mass reformatting instead of using pre-commit. And some of these env var settings seem unnecessary and look like personal preference. But again, I have no idea how these scripts work because I'm not a nixOS user.

**Comment by uklotzde** [comment]
I am also uncomfortable with mixing platform dependent stuff besides generic Linux/Windows/macOS support into projects. If this is necessary then it is a design flaw of NixOS.

**Comment by JoergAtGithub** [comment]
The fact, that two independent users created PRs for this, shows me, that there's a need for this.

**Comment by Holzhaus** [comment]
> The fact, that two independent users created PRs for this, shows me, that there's a need for this.

Because the current state is broken, which is arguably worse than having no nixOS scripts at all and sheds a bad light on Mixxx. When we have scripts or documentation for something, it's reasonable to expect that it works.

I'd still prefer to remove them completely from our source tree and let the nixOS users maintain them in a more appropriate location.

File: shell.nix
Comment (poelzi) [reply shell.nix:56]: when i started this patch, the pre-commit shipped with 19.xx was to old for our configuration and commit testing was not possible. I think I had problems backporting to a derivate so I installed it ugly in the shell. removed now.

**Comment by poelzi** [comment]
@Holzhaus so, should I prepare a patch removing everything packaging related then ? no more packages for windows, debian, rpm etc. Because I can't see your point. Compared to all the other options, nix collects its garbage and does not force me to download and install countless other packages just to compile some peace of software, maybe even just once. Just compare the wiki with instructions how to setup a development environment.

`rg qt5keychain-dev
tools/debian_buildenv.sh
85:            qt5keychain-dev \

packaging/debian/control.in
21:               qt5keychain-dev,
`
Why not start removing that garbage first. The one nix file is able to setup fully working dev environment with one command. A fully build package with another, hey, you can even select the compiler and addon with simple arguments. Compared to that shipped package systems, the one nix file is able to reproduce binary identical builds. If you looked arround lately, all kind of projects ship with nix files because it makes lives easier. You don't have to run nixos to run the nix package manager.

**Comment by uklotzde** [comment]
We already discussed that Debian packaging does not belong into the main repo, as for any other distribution out there. No need to argue about that.

**Comment by daschuer** [comment]
I think we really need to try to keep this discussing objective. @Holzhaus is trying hard to get the release out and has considered this PR as a candidate to be included. Unfortunately, this PR has open review comments and is quite big, mixing nix related changes with changed affecting all build systems, which makes it unhandy and adds a regression risk. We need here a balance between the risk for all compared to the benefit for some users.

@poelzi do you see a chance to find time to target the review comments soon? That would be grate. It would be also helpful if you split out the uncritical parts. I think that they can be instantly merged.

In general I am on the pro side of merging packageing related PRs as long they are actively maintained. Such PRs are reducing the risk that package maintainers do something wrong. Maintaining them in place with the code is IMHO better then on a wiki page which is out of sight when related changes are introduced.

**Comment by poelzi** [comment]
I want to emphasis again, that this nix file is not meant to be included in nixpkgs, I'm pretty sure they will update to the next stable release when published. For example, dependencies like djinterop do not belong into the mixxx definition, but at their proper place in the tree. The developer support in there is also not important, because you are not meant to run precommit hooks there. This nix file is meant for developers, not users. You can use it to build a package for your PC, because you obviously want to eat your own dogfood (branch) next to other branches. For my day to day version I'm building my package exactly the same way I'm building my development branches as well.

If you can make it easier for somebody to contribute by lowering the burden for getting a dev environment running, it is the right thing to do.

**Comment by flosse** [comment]
Hi guys, I'm a NixOS user and I'm really happy that there is a nix file for development **BUT** I also agree that NixOS related stuff doesn't necessarily needs to be maintained within the upstream project.
So I can imagine two solutions:

1. Remove all NixOS related stuff and create a `mixxx-for-nixos-dev` git repository **AND** add a link to the repo in the README or mixxx website

OR

2. A person that is a **member of the core team** promises to constantly maintain the nix files.

---

BTW:
I [enjoy using Mixxx](https://soundcloud.com/munkelbeatz/live-villa-kunterbunt-2021-03-13 ) on NixOS + Denon MC7000

**Review by uklotzde — CHANGES_REQUESTED** [review]
We didn't agree on adding a NixOS CI build.

**Comment by poelzi** [comment]
@uklotzde it broke because there was no ci for nix. I'm happy with maintaining the nix support, but if I have to remove the CI I will just close this PR and move on. This Issue is becoming jet another pain point

**Comment by uklotzde** [comment]
Maintaining this additional CI build will be a paint point for everyone else.

The only circumstances under which I would accept it is if we are allowed to disable it as soon as anything breaks. But then what is a contributor supposed to do if the NixOS CI build breaks in a PR? Shall we reject the PR as long as the CI build fails?

**Comment by uklotzde** [comment]
If the RPM Fusion build breaks I either create a custom, temporary patch and/or open upstream PRs. The packaging is separated from the Mixxx source repo. We don't need to bother everyone else with Fedora dependencies, RPM packaging, or support for unofficial targets like armv7hl/aarch64/power64.

Yes, this requires some extra work. But it would be unfair to put this burden on everyone else that doesn't care about Fedora or NixOS.

**Comment by poelzi** [comment]
@uklotzde the nix builds can now fail and will not cause the pipline to fail.

If that would be true, #3572 could be merged already....

**Comment by uklotzde** [comment]
@poelzi #3572 still has multiple unresolved issues. I don't get why you mention this PR here?

File: CMakeLists.txt
Comment (daschuer) [inline CMakeLists.txt:62]: Please name the glibc version. The issue starts with glibc-2.31
https://github.com/bminor/glibc/commit/7bdb921d70bf9f93948e2e311fef9ef439314e41

On Ubuntu we have not issue because Ubuntu Bionic uses glibc-2.27 and LVVM 9 
Ubuntu Focal has 2.31 but also LVVM 10.

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by uklotzde** [comment]
Wait, we are not only adding 1 but 2 NixOS builds? How far do we want to stress the CI builds? Sorry, but I disagree.

Opinions?

**Comment by uklotzde** [comment]
Just for my understanding: Why do we need to pull the NixOS artifacts into the Mixxx source repo and extend our CI builds instead of hosting the configuration files in [nixpkgs](https://github.com/NixOS/nixpkgs/tree/master/pkgs/applications/audio/mixxx)? There is even a dedicated build system for NixOS packages.

**Comment by Holzhaus** [comment]
How do we continue here? We shouldn't tag 2.3.0 with a broken nixOS config. I think we have 4 alternatives:
1. Merge this PR including CI builds
2. Merge this PR excluding CI builds
3. Merge https://github.com/mixxxdj/mixxx/pull/3795 instead of this PR
4. Merge neither and remove the nixOS config from our source tree

Opinions?

**Comment by JoergAtGithub** [comment]
I'm in favour of #1, because the CI build with Clang compiler ensures compatibility, not only for NIX.

**Comment by Holzhaus** [comment]
> I'm in favour of #1, because the CI build with Clang compiler ensures compatibility, not only for NIX.

We already ensure clang compatibility with he clang-tidy, the clazy and the macOS CI builds.

**Comment by uklotzde** [comment]
The CI builds are not needed. NixOS users will notice any breakage when building locally and could then file a PR. Why should we maintain CI builds that affect only very few users, are unmaintainable for most of us, and are not needed for any official release? We also don't host CI builds for other Linux distributions.

**Comment by uklotzde** [comment]
> > I'm in favour of #1, because the CI build with Clang compiler ensures compatibility, not only for NIX.
> 
> We already ensure clang compatibility with the clang-tidy, clazy and macOS CI builds.

Correct, we already have one CI build for each compiler suite (GCC, Clang, MSVC). We cannot test all possible combinations including different versions, platforms, CPU architectures, ...

**Comment by Holzhaus** [comment]
I'd prefer 4, 3 or 2, in that order. If we chose 1, all PRs with failing nixOS builds are essentially blocked until @poelzi reviews then and contributes a fix. Or we ignore if the nixOS CI fails, which makes them kinda pointless.

**Comment by daschuer** [comment]
I vote for 2 or 3. If we decide for 2, I would prefer to cut out unrelated changes into single PRs.

**Comment by daschuer** [comment]
Can we close this stale PR now? Our main build tool is cmake and we don't have the power to maintain another build script of 397 lines. 
See the discussion in https://github.com/mixxxdj/mixxx/pull/3795#issuecomment-1015186782 for the option striping it down to an environment setup only tool.

**Comment by cr7pt0gr4ph7** [comment]
Currently going through old & stale PRs in an effort to reduce the number of open PRs.
I'd argue that if this build script update wasn't relevant two years ago, it very much isn't relevant now, and can be closed.

**Comment by daschuer** [comment]
Yes, let's close this and than also decide for
https://github.com/mixxxdj/mixxx/pull/3795
as well.

**Comment by cr7pt0gr4ph7** [comment]
> Yes, let's close this and than also decide for #3795 as well.

#4632 would also be a candidate for closing then.

**Pull request closed without merging**
